### PR TITLE
Extract Field Action: Make it possible to set if the extracted field should be in the body or in a particular header

### LINKED ISSUE #677: Extract Field Action: Make it possible to set if the extracted field should be in the body or in a particular header

### PR DESCRIPTION
**Pull request by oscerd**
Fixes #677

### CODE REVIEW

**Comment by oscerd** [comment]
Related to #673

File: kamelets/extract-field-action.kamelet.yaml [hunk 1]
```diff
@@ -29,14 +29,30 @@ metadata:
 spec:
   definition:
     title: "Extract Field Action"
-    description: "Extract a field from the body"
+    description: |-
+      Extract a field from the message body.
+
+      The extract field action expected an application/json content type.
```
Comment (ReggieCarey) [inline kamelets/extract-field-action.kamelet.yaml:35]: The extract field action expects an application/json content type.

File: kamelets/extract-field-action.kamelet.yaml [hunk 2]
```diff
@@ -29,14 +29,30 @@ metadata:
 spec:
   definition:
     title: "Extract Field Action"
-    description: "Extract a field from the body"
+    description: |-
+      Extract a field from the message body.
+
+      The extract field action expected an application/json content type. 
+
+      The field parameter allows to specify which field of the json the user wants to extract. By default the message body will be overriden with the extracted field.
```
Comment (ReggieCarey) [inline kamelets/extract-field-action.kamelet.yaml:37]: The field parameter allows you to ...

File: library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/transform/ExtractField.java [hunk 3]
```diff
@@ -27,11 +27,16 @@
...
+    public void process(@ExchangeProperty("field") String field, @ExchangeProperty("headerOutput") boolean headerOutput, Exchange ex) {
+        final String EXTRACTED_FIELD_HEADER = "CamelKameletsExtractFieldName";
         ObjectMapper mapper = new ObjectMapper();
         JsonNode jsonNodeBody = ex.getMessage().getBody(JsonNode.class);
         Map<Object, Object> body = mapper.convertValue(jsonNodeBody, new TypeReference<Map<Object, Object>>(){});
-        ex.getMessage().setBody(body.get(field));
+        if (!headerOutput) {
+            ex.getMessage().setBody(body.get(field));
+        } else {
+            ex.getMessage().setHeader(EXTRACTED_FIELD_HEADER, body.get(field));
```
Comment (ReggieCarey) [inline library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/transform/ExtractField.java:38]: As per my comments in the issue, maybe this is better in line 38?

ex.getMessage.setHeader(sanitized(headerOutput), body.get(field));

Where ```String sanitized(String headerOutput)``` does its best to ensure a valid headerName that does not collide with existing headers.

Comment (oscerd) [reply library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/transform/ExtractField.java:38]: Header output is a boolean. The header name is static and fixed. So, no.

**Comment by oscerd** [comment]
I'm starting to be a bit bored by the way you're going through this stuff. Maybe someone else will want to help you. Have a nice weekend

File: library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/transform/ExtractField.java (hunk 3 above)
Comment (oscerd) [reply library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/transform/ExtractField.java:38]: Also this is a bean that should be simple as simple. I troducing more logic to check stuff like headers existence is too much, there is for example the has header action or you can write your own kamelet action doing all the logic you want. But I'd probably leave the beans simplest as possible.

**Comment by ReggieCarey** [comment]
I don't understand - you seem to think I'm attacking you.  Nothing could be farther from the truth.

You want community involvement but every time a question, suggestion or involvement is offered, you're offended.

If you don't like my suggestions/questions that's fine.  But I'm trying to keep this conversation from devolving into something inappropriate for public discourse on an open source project.  I truly am interested in a dialog about improving this for myself and others.

The issue tool is intended to allow public discussion and debate as well as to get ideas circulated that improves the product for all.  Some ideas are great and some not so great. Maybe my suggestions don't align with the intended product direction. 

Remember that the only people willing to spend any time at all writing issues and commenting have a vested interest in seeing this project succeed for everyone.  Attacking people that raise issues with what has been produced as open source is counter productive to the open source ideals and serves only to push people away.

If this is not the venue by which we community members make suggestions, then what is?

**Comment by oscerd** [comment]
I don't discuss the where, or the issue, I'm discussing the way. But it might be it's just me.
You and anyone else are always welcome to discuss and open issues. Probably i was already upset by all the reksted discussion from yesterday and I'm still upset, but not with you in particular. 

By the way let's have another discussiin after the weekend. Thanks for the suggestions and for the help.

**Comment by oscerd** [comment]
I found some time to answer now with a bit more lucidity than yesterday. I answered on the issue related to ES Index Sink Kamelet. Hope this clarifies some stuff, but lets continue the discussion there. Cheers.

**Pull request merged**